text/0000-get-mut-refs.md

```diff
@@ -0,0 +1,249 @@
+- Feature Name: get-mut-refs
+- Start Date: 2021-03-11
+- RFC PR: TBD
+- Rust Issue: TBD
+
+# Definitions:
+
+- "child place(s)": similar in concept to a struct field but opaque to the borrow checker, examples being an array index or a hash map's entry.
+- "dependant trait": a trait which abstracts over some requirement of another trait
+
+# Summary
+[summary]: #summary
+
+Add a new trait and dependant unsafe trait which together abstract over the retrieval of multiple `&mut` to child places.
+The normal trait would be added as part of the v1 prelude so that these functions are easy to find and use.
+
+Example: (assume the trait function is called `get_mut_refs`)
+```rust
+let place1: usize = ...;
+let place2: usize = ...; // where place1 != place2
+
+let mut x = [0, 1, 2, 3, 4];
+
+if let Some([x1, x2]) = x.get_mut_refs([place1, place2]) {
+  swap(x1, x2);
+}
+
+dgb!{x};
+```
+
+# Motivation
+[motivation]: #motivation
+
+This is currently something that is not possible to do in safe code in rust because the borrow checker doesn't know how to deal with indexing or other forms of child places.
+This mechanism would provide a standard way for types to implement the retrieval of multiple `&mut` to child places.
+Because this exposes a safe API it can be used by everyone easily and thus reduces the difficulty of wanting to get multiple `&mut` into data structures at arbitrary locations.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Two new traits are added the `core` called `GetMutRefs` and `GetMutRefsRawEntry`.
+It is not needed to import `GetMutRefs` as it is part of the prelude.
+Unless you are implementing `GetMutRefs` for your own type the it is not needed to worry about `GetMutRefsRawEntry`.
+
+Normally, when rust code uses the `IndexMut` or `BorrowMut` trait it borrows the whole owned value and returns a single `&mut`.
+However, for some data structures, such as arrays and vectors, it is possible and sometimes desirable to get more than one `&mut` out.
+This is because these data structures represent a managed collection of values which are logically (and from a memory safety point of view) disjoint from each other.
+
+This trait is the common way for a data structure to expose this disjoint nature of its child places.
+When the `get_mut_refs` method is called it checks that all keys are valid and that none of the resulting `&mut` would violate the `&mut` contracts.
+If either of those are false then `None` is returned.
+
+## Examples:
+
+### Individual Entries:
+
+```rust
+let mut x = [0, 1, 2, 3, 4];
+
+if let Some([x1, x2]) = x.get_mut_refs([0, 4]) {
+  swap(x1, x2);
+}
+
+dgb!{x}; // [4, 1, 2, 3, 0]
```

Comment:
Typo: That should be `dbg!{x};`

Reply:
Oops thanks.

Reply:
This typo is also in the first code block.